notify/jira/jira.go

```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		return issue{}, fmt.Errorf("template error: %w", err)
+	}
+
+	if priority != "" {
+		requestBody.Fields.Priority = &idNameValue{Name: priority}
+	}
+
+	sort.Strings(requestBody.Fields.Labels)
+
+	return requestBody, nil
+}
+
+func (n *Notifier) searchExistingIssue(key notify.Key, status model.AlertStatus) (*issue, bool, error) {
+	jql := strings.Builder{}
+
+	if n.conf.WontFixResolution != "" {
+		jql.WriteString(fmt.Sprintf(`resolution != %q and `, n.conf.WontFixResolution))
+	}
+
+	// if the alert is firing, do not search for closed issues unless reopen transition is defined.
+	if n.conf.ReopenTransition == "" {
+		if status != model.AlertResolved {
+			jql.WriteString(`statusCategory != Done and `)
+		}
+	} else {
+		reopenDuration := int64(time.Duration(n.conf.ReopenDuration).Minutes())
+		if reopenDuration != 0 {
+			jql.WriteString(fmt.Sprintf(`(resolutiondate is EMPTY OR resolutiondate >= -%dm) and `, reopenDuration))
+		}
+	}
+
+	alertLabel := fmt.Sprintf("ALERT{%s}", key.Hash())
+	jql.WriteString(fmt.Sprintf(`project=%q and labels=%q order by status ASC,resolutiondate DESC`, n.conf.Project, alertLabel))
+
+	requestBody := issueSearch{}
+	requestBody.Jql = jql.String()
+	requestBody.MaxResults = 2
+	requestBody.Fields = []string{"status"}
+	requestBody.Expand = []string{}
+
+	level.Debug(n.logger).Log("msg", "search for recent issues", "alert", key.String(), "jql", jql.String())
+
+	responseBody, shouldRetry, err := n.doAPIRequest(http.MethodPost, "search", requestBody)
+	if err != nil {
+		return nil, shouldRetry, err
+	}
+
+	var issueSearchResult issueSearchResult
+	err = json.Unmarshal(responseBody, &issueSearchResult)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if issueSearchResult.Total == 0 {
+		level.Debug(n.logger).Log("msg", "found no existing issue", "alert", key.String())
+		return nil, false, nil
+	}
+
+	if issueSearchResult.Total > 1 {
+		level.Warn(n.logger).Log("msg", "more than one issue matched, selecting the most recently resolved", "alert", key.String(), "selected", issueSearchResult.Issues[0].Key)
+	}
+
+	return &issueSearchResult.Issues[0], false, nil
+}
+
+func (n *Notifier) getIssueTransitionByName(issueKey, transitionName string) (string, bool, error) {
+	path := fmt.Sprintf("issue/%s/transitions", issueKey)
+
+	responseBody, shouldRetry, err := n.doAPIRequest(http.MethodGet, path, nil)
+	if err != nil {
+		return "", shouldRetry, err
+	}
+
+	var issueTransitions issueTransitions
+	err = json.Unmarshal(responseBody, &issueTransitions)
+	if err != nil {
+		return "", false, err
+	}
+
+	for _, issueTransition := range issueTransitions.Transitions {
+		if issueTransition.Name == transitionName {
+			return issueTransition.ID, false, nil
+		}
+	}
+
+	return "", false, fmt.Errorf("can't find transition %s for issue %s", transitionName, issueKey)
+}
+
+func (n *Notifier) TransitionIssue(key notify.Key, issueKey, transitionName string) (bool, error) {
```

Comment:
This comment is more of a curiosity. I was wondering why this function starts with a capital letter. In my understanding, the functions (Golang) with the first capital letter are normally used when they're referenced from outside the package, which makes sense for the New() function, but TransitionIssue() seemed out of the norm.

Reply:
Good catch! I guess its an leftover since I do some copy paste from jiralert